pkg/scheduler/algorithm/predicates/metadata_test.go

```diff
@@ -507,3 +507,235 @@ func TestPredicateMetadata_ShallowCopy(t *testing.T) {
 		t.Errorf("Copy is not equal to source!")
 	}
 }
+
+// TestGetPodsMatchingAffinity_AntiAffinity tests againt method getPodsMatchingAffinity
+// on Anti Affinity cases
+func TestGetPodsMatchingAffinity_AntiAffinity(t *testing.T) {
+	newPodAffinityTerms := func(keys []string) []v1.PodAffinityTerm {
+		var terms []v1.PodAffinityTerm
+		for _, key := range keys {
+			terms = append(terms, v1.PodAffinityTerm{
+				LabelSelector: &metav1.LabelSelector{
+					MatchExpressions: []metav1.LabelSelectorRequirement{
+						{
+							Key:      key,
+							Operator: metav1.LabelSelectorOpExists,
+						},
+					},
+				},
+				TopologyKey: "hostname",
+			})
+		}
+		return terms
+	}
+	tests := []struct {
+		name                 string
+		existingPods         []*v1.Pod
+		nodes                []*v1.Node
+		pod                  *v1.Pod
+		wantAffinityPods     map[string][]*v1.Pod
+		wantAntiAffinityPods map[string][]*v1.Pod
+		wantErr              bool
+	}{
+		{
+			name: "nil test",
+			nodes: []*v1.Node{
+				{ObjectMeta: metav1.ObjectMeta{Name: "nodeA", Labels: map[string]string{"hostname": "nodeA"}}},
+			},
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{Name: "aaa-normal", Labels: map[string]string{"aaa": ""}},
+			},
+			wantAffinityPods:     nil,
+			wantAntiAffinityPods: nil,
+		},
+		{
+			name: "non-affinity incoming pod causes a no-op", // the semmetric verification is covered elsewhere
+			existingPods: []*v1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "aaa-anti", Labels: map[string]string{"aaa": ""}},
+					Spec: v1.PodSpec{
+						Affinity: &v1.Affinity{
+							PodAntiAffinity: &v1.PodAntiAffinity{
+								RequiredDuringSchedulingIgnoredDuringExecution: newPodAffinityTerms([]string{"aaa"}),
+							},
+						},
+						NodeName: "nodeA",
+					},
+				},
+			},
+			nodes: []*v1.Node{
+				{ObjectMeta: metav1.ObjectMeta{Name: "nodeA", Labels: map[string]string{"hostname": "nodeA"}}},
+			},
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{Name: "aaa-normal", Labels: map[string]string{"aaa": ""}},
+			},
+			wantAffinityPods:     nil,
+			wantAntiAffinityPods: nil,
+		},
+		{
+			name: "no pod has label violates with incoming pod's anti-affinity",
+			existingPods: []*v1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "bbb-normal", Labels: map[string]string{"bbb": ""}},
+					Spec:       v1.PodSpec{NodeName: "nodeA"},
+				},
+			},
+			nodes: []*v1.Node{
+				{ObjectMeta: metav1.ObjectMeta{Name: "nodeA", Labels: map[string]string{"hostname": "nodeA"}}},
+			},
+			pod: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{Name: "aaa-anti", Labels: map[string]string{"aaa": ""}},
+				Spec: v1.PodSpec{
+					Affinity: &v1.Affinity{
+						PodAntiAffinity: &v1.PodAntiAffinity{
+							RequiredDuringSchedulingIgnoredDuringExecution: newPodAffinityTerms([]string{"aaa"}),
+						},
+					},
+				},
+			},
+			wantAffinityPods:     make(map[string][]*v1.Pod),
```

Comment:
You used nil above, should be consistent :D 

btw, I think nil is fine.

Reply:
Good catch, lol...

`getPodsMatchingAffinity` simply return nil if it doesn't have affinity/antiaffinity:

https://github.com/kubernetes/kubernetes/blob/058b26f38e49b64faad4a7a346b4616d0e9b57a0/pkg/scheduler/algorithm/predicates/metadata.go#L376-L382

But Mr. `reflect.DeepEqual` is unhappy to treat "nil" as equality of a "nil map" - so in some cases, I put nil, while some I put nil map there to make it happy...

BTW: `apiequality.Semantic.DeepEqual` is considerate to say yes to "nil equals to nil map". But maybe not worth introducing an extra package dependency.

Reply:
Ah I see, maybe just return empty object every time is enough.

Reply:
Hmm... Have to rebase code as PR #67788 refactored the data structure.

Not it's all consistent. PTAL.